### DISCUSSION TITLE
A better way to trigger suspend when the user is idle?

### DISCUSSION THREAD

**Discussion by rlue** [i3/i3#5030] [Q&A] [answered]
I want to put my system in suspend mode when I'm afk. I am aware of utilities like xautolock and xidlehook, but the way that they define/detect idleness is not ideal for me.

#### xautolock

AFAICT, xautolock just watches for keyboard/mouse input. This is problematic when, say, you're watching a video on YouTube—you're not using your keyboard or mouse, but you don't want to trigger power management features, either.

A [common remedy to this is to use the `-corners` option](https://www.reddit.com/r/i3wm/comments/ak8fjy/how_do_you_guys_suspend_xautolocki3lock_when/), but that requires learning a new habit, which I'd like to avoid.

#### xidlehook

xidlehook takes this a step further by offering options for `--not-when-fullscreen` and `--not-when-audio`. These are a step in the right direction (and for some people, maybe even better than how X handles DPMS—see below), but for me, it's still a crude approximation of real idle detection:

1. "fullscreen" is not quite the right criterion. If a video is playing but not full-screen, I want idle detection inhibited; if it's full-screen but paused, I want idle detection _on._
2. `--not-when-audio` can get totally broken by bugs like [this one](https://bugzilla.mozilla.org/show_bug.cgi?id=1743834), where firefox just forgets to close the audio stream, even though nothing's playing. In cases like this, xidlehook won't trigger idle actions until the application is closed (which, in the case of my browser, is never).
3. It's not compatible with `caffeine` (also the case with xautolock).

### Then how do you want it done?

The `xset` utility allows you to modify some basic preferences in X, including screensaver / display shutoff timeouts:

```sh
$ xset s off     # disable the screen saver
$ xset dpms 300  # enter DPMS (Display Power Management System) standby after 5 minutes
```

The behavior of these timeouts turns out to be my preferred idle detection scheme. It is **context-aware**—that is, it is inhibited when:

1. watching a video in VLC or mpv, regardless of whether it's full-screen, but **only when it's actually playing**
2. videoconferencing or watching a video in Firefox, but **only in a foreground tab**
3. using `caffeine`

(Sadly, this context-awareness depends on implementation in each application, and the Zoom guys dropped the ball on that one a long time ago and never bothered chasing after it. No matter! I can wrap zoom in a little caffeine shim for that.)

### What I want

I want to get basic power management working with i3, which means two things: 1) display shutoff and 2) system suspend. Moreover, I want them to follow the same rules about what "idle" means, and I really want that definition to be the one the X server uses for DPMS.

* ✅ **Display shutoff:** Luckily, xset already provides everything I need here.
* ❌ **System suspend:** I have tried:
  * power managers like [powerdevil](https://github.com/KDE/powerdevil) and [powerkit](https://github.com/rodlie/powerkit)—sadly, neither one inhibits idle at the right times with i3 (system went down for suspend in the middle of watching a video)
  * `/etc/systemd/logind.conf` (specifically, the `IdleAction` and `IdleActionSec` options)—for some reason, the system doesn't go down for suspend at all with this approach (according to [this thread on the ArchLinux forums](https://bbs.archlinux.org/viewtopic.php?id=252877), "i3 doesn't actually set the Idle Information that would make this work anyway"?)

Does anyone have a better solution? Surely, there must be a way. 🙏🙏🙏

**Comment by vincentbernat** [comment, marked as answer]
For the idle hint, you can use `xss-lock` that will manage it through the MIT screen saver extension. Moreover, it's a good way to reliably start your screen locker.

**Comment by Airblader** [comment]
> If a video is playing but not full-screen, I want idle detection inhibited

That sounds like a perfect foot gun. Your browser is open, some ad video starts playing and your machine never locks.

**Reply by rlue** [reply]
1. I've got adblock, thanks.
2. This only applies to foreground tabs.
3. How is that worse than all the other ways that other idle detectors can get it wrong?

No disrespect—thank you for your work on i3-gaps.
